clients/venice-admin-tool/src/main/java/com/linkedin/venice/ZkCopier.java

```diff
@@ -110,6 +159,54 @@ static TreeNode buildRequiredPathsTree(Set<String> clusterNames, String basePath
     return root;
   }
 
+  /**
+   * Build a sorted list of Venice-specific paths in the ZK client filtered by 1)base path, 2)cluster names, and 3)required cluster ZK paths.
+   * @return a sorted list of Venice-specific paths
+   */
+  static List<String> getZkClientPathsList(ZkClient zkClient, Set<String> clusterNames, String basePath) {
+    List<String> paths = new ArrayList<>();
+    StringBuilder sb = new StringBuilder(basePath);
+    paths.add(sb.toString());
+    String storeConfigsPath = sb.append("/").append(STORE_CONFIGS).toString();
+    if (zkClient.exists(storeConfigsPath)) {
+      paths.add(storeConfigsPath);
+      getZkClientPathsListHelper(zkClient, paths, sb);
+    }
+    sb.delete(sb.lastIndexOf("/" + STORE_CONFIGS), sb.length());
+    for (String cluster: clusterNames) {
+      String clusterPath = sb.append("/").append(cluster).toString();
+      if (zkClient.exists(clusterPath)) {
+        paths.add(clusterPath);
+        for (String venicePath: CLUSTER_ZK_PATHS) {
+          String clusterVenicePath = sb.append("/").append(venicePath).toString();
+          if (zkClient.exists(clusterVenicePath)) {
+            paths.add(clusterVenicePath);
+            getZkClientPathsListHelper(zkClient, paths, sb);
+          }
+          sb.delete(sb.lastIndexOf("/" + venicePath), sb.length());
+        }
+      }
+      sb.delete(sb.lastIndexOf("/" + cluster), sb.length());
+    }
+    Collections.sort(paths);
+    return paths;
+  }
```

Comment:
A couple of comments in this function:
1. Can we avoid using a single StringBuilder to orchestrate this whole logic, and also do away with the `deletes`? It is hard to make a mental model of things if there is only one object that get mutated quite frequently.
2. Instead of checking `zkClient.exists` multiple times, each of which would incur a remote query to Zk, we should do `zkClient.getChildren`, and then from that list, pick and choose the stuff that we need.

Reply:
Previous iteration was better https://github.com/linkedin/venice/pull/1056#discussion_r1683600582